src/compiler/parser.ts

```diff
@@ -2226,16 +2257,34 @@ module ts {
             }
         }
 
+        function setContextParameters(explicitFlags: ParserContextFlags, ambientFlags: ParserContextFlags) {
+            // setContextParameters atomically returns the value of contextFlags at the time the function was 
+            // called, and updates contextFlags in the following fashion:
+            //
+            // - Any parser-generated flags not in the set of ambient flags are cleared from contextFlags
+            // - Explicit flags are set on contextFlags
+
+            var previousContextFlags = contextFlags;
+            contextFlags = (contextFlags & (ambientFlags | ~ParserContextFlags.ContextParameterFlags)) | explicitFlags;

```

Comment:
I understand now. This is very ES6-y. I prefer the model where ambient flags are passed by just doing nothing, and flags are explicitly turned on and off. I know that what you've done here is more to the grammar, but for some reason I find it less intuitive.